### PR TITLE
Product category api

### PR DESCRIPTION
**Pull request by AyalaSprei**

### CODE REVIEW

File: src/main/java/com/yellow/ordermanageryellow/Dao/ProductCategoryRepository.java [hunk 1]
```diff
@@ -1,8 +1,20 @@
...
+import java.util.Optional;
+
 public interface ProductCategoryRepository extends MongoRepository<ProductCategory, String> {
+    List<ProductCategory> findAll();
+    ProductCategory save(ProductCategory productCategory);
+    // Delete
+    void deleteById(String categoryId);
+
+    // Edit
+    //ProductCategory update(ProductCategory productCategory);
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Dao/ProductCategoryRepository.java:16]: please remove commented lines

File: src/main/java/com/yellow/ordermanageryellow/Dao/ProductCategoryRepository.java [hunk 2]
```diff
@@ -1,8 +1,20 @@
...
 import com.yellow.ordermanageryellow.model.ProductCategory;
-import com.yellow.ordermanageryellow.model.Users;
 import org.springframework.data.mongodb.repository.MongoRepository;
 
+import java.util.List;
+import java.util.Optional;
+
 public interface ProductCategoryRepository extends MongoRepository<ProductCategory, String> {
+    List<ProductCategory> findAll();
+    ProductCategory save(ProductCategory productCategory);
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Dao/ProductCategoryRepository.java:11]: no need to implement save, there is already insert method in MongoRepository

File: src/main/java/com/yellow/ordermanageryellow/Dao/ProductCategoryRepository.java [hunk 3]
```diff
@@ -1,8 +1,20 @@
...
 
 import com.yellow.ordermanageryellow.model.ProductCategory;
-import com.yellow.ordermanageryellow.model.Users;
 import org.springframework.data.mongodb.repository.MongoRepository;
 
+import java.util.List;
+import java.util.Optional;
+
 public interface ProductCategoryRepository extends MongoRepository<ProductCategory, String> {
+    List<ProductCategory> findAll();
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Dao/ProductCategoryRepository.java:10]: no need implement it, already exists in super

File: src/main/java/com/yellow/ordermanageryellow/Dao/ProductCategoryRepository.java [hunk 4]
```diff
@@ -1,8 +1,20 @@
...
+    ProductCategory save(ProductCategory productCategory);
+    // Delete
+    void deleteById(String categoryId);
+
+    // Edit
+    //ProductCategory update(ProductCategory productCategory);
+
+    // Retrieve all
+    Optional<ProductCategory> findById(String id);
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Dao/ProductCategoryRepository.java:19]: as above no need to implement it

File: src/main/java/com/yellow/ordermanageryellow/Service/ProductCategoryService.java [hunk 5]
```diff
@@ -4,18 +4,58 @@
...
-    @Override
-    public void run(String... args) {
-        ProductCategory ProductCategory = new ProductCategory("12");
-        ProductCategoryRepository.save(ProductCategory);
+    //@Override
+    //public void run(String... args) {
+       // ProductCategory ProductCategory = new ProductCategory("12");
+    //  ProductCategoryRepository.save(ProductCategory);
+   // }
+    public ResponseEntity<List<ProductCategory>> findAll(){
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Service/ProductCategoryService.java:25]: all methodes in service should return the object itself - List<ProductCategory>
ResponseEntity<List<ProductCategory>> - should return in controller itself

File: src/main/java/com/yellow/ordermanageryellow/api/ProductCategoryController.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+import com.yellow.ordermanageryellow.model.ProductCategory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.List;
+
+@RestController
+@RequestMapping("/categories")
+public class ProductCategoryController {
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/api/ProductCategoryController.java:12]: please add exception handling

File: src/main/java/com/yellow/ordermanageryellow/Service/ProductCategoryService.java [hunk 7]
```diff
@@ -4,18 +4,58 @@
...
+        List<ProductCategory> categories = this.ProductCategoryRepository.findAll();
+        return new ResponseEntity<>(categories, HttpStatus.OK);    }
+
+    public ResponseEntity<String> insert(ProductCategory newCategory) {
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Service/ProductCategoryService.java:29]: you should save the product category with comapnay from token

**Review by moriyahazan — CHANGES_REQUESTED** [review]
@AyalaSprei  good job
please over me comments and update accordantly

File: src/main/java/com/yellow/ordermanageryellow/Service/ProductCategoryService.java (hunk 7 above)
Comment (AyalaSprei) [reply src/main/java/com/yellow/ordermanageryellow/Service/ProductCategoryService.java]: I wrote you an email, with some questions and comments
For now I did what I thought
so you can see here and read my email...

File: src/main/java/Exceptions/ObjectAlreadyExistException.java [hunk 8]
```diff
@@ -0,0 +1,7 @@
+package Exceptions;
+
+public class ObjectAlreadyExistException extends RuntimeException{
```
Comment (moriyahazan) [inline src/main/java/Exceptions/ObjectAlreadyExistException.java:3]: Should extend from Exception not RuntimeException

File: src/main/java/com/yellow/ordermanageryellow/Service/ProductCategoryService.java [hunk 9]
```diff
@@ -1,21 +1,44 @@
...
-    @Override
-    public void run(String... args) {
-        ProductCategory ProductCategory = new ProductCategory("12");
-        ProductCategoryRepository.save(ProductCategory);
+
+    public List<ProductCategory> findAll(){
+            return this.ProductCategoryRepository.findAll();
+    }
+    public ProductCategory insert(String name, String description) {
+        if(this.ProductCategoryRepository.existsByName((name))==true)
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Service/ProductCategoryService.java:26]: no need ==true

File: src/main/java/com/yellow/ordermanageryellow/Service/ProductCategoryService.java [hunk 10]
```diff
@@ -1,21 +1,44 @@
...
+import org.springframework.web.bind.annotation.PathVariable;
+
+import java.util.List;
+import java.util.NoSuchElementException;
+import java.util.Optional;
 
 @Service
-public class ProductCategoryService implements CommandLineRunner {
+public class ProductCategoryService  {
     private final ProductCategoryRepository ProductCategoryRepository;
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Service/ProductCategoryService.java:16]: the variable should start with a lower latter 
    private final ProductCategoryRepository productCategoryRepository;

File: src/main/java/com/yellow/ordermanageryellow/Service/ProductCategoryService.java [hunk 11]
```diff
@@ -1,21 +1,44 @@
...
     }
-    @Override
-    public void run(String... args) {
-        ProductCategory ProductCategory = new ProductCategory("12");
-        ProductCategoryRepository.save(ProductCategory);
+
+    public List<ProductCategory> findAll(){
+            return this.ProductCategoryRepository.findAll();
+    }
+    public ProductCategory insert(String name, String description) {
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Service/ProductCategoryService.java:25]: The parameter should be Object ProductCategory

File: src/main/java/com/yellow/ordermanageryellow/Service/ProductCategoryService.java [hunk 12]
```diff
@@ -1,21 +1,44 @@
...
+        ProductCategory newCategory= new ProductCategory();
+        newCategory.setName(name);
+        newCategory.setDesc(description);
+        return this.ProductCategoryRepository.save(newCategory);
+    }
+        public void delete(String categoryId) {
+            this.ProductCategoryRepository.deleteById(categoryId);
+        }
+
+    public ProductCategory update(String categoryId, String name, String description){
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Service/ProductCategoryService.java:37]: Parameter should be ProductCategory  not each value separately

File: src/main/java/com/yellow/ordermanageryellow/api/ProductCategoryController.java [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+            return new ResponseEntity<>(categories, HttpStatus.OK);
+        }
+        catch (Exception ex){
+            return new ResponseEntity(HttpStatus.INTERNAL_SERVER_ERROR);
+        }
+    }
+
+    @PostMapping("/{name}/{description}")
+    public ResponseEntity createCategory(@PathVariable String name,@PathVariable String description) {
+        ProductCategory createdCategory;
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/api/ProductCategoryController.java:40]: Please get object like i wrote in service

File: src/main/java/com/yellow/ordermanageryellow/api/ProductCategoryController.java [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+        catch(EmptyResultDataAccessException ex){
+            return new ResponseEntity(HttpStatus.NOT_FOUND);
+        }
+       catch (Exception ex){
+           return new ResponseEntity(HttpStatus.INTERNAL_SERVER_ERROR);
+       }
+    }
+
+    @PutMapping("/{categoryId}/{name}/{description}")
+    public ResponseEntity updateCategory(@PathVariable String categoryId,@PathVariable String name, @PathVariable String description) {
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/api/ProductCategoryController.java:68]: Get object

File: src/main/java/com/yellow/ordermanageryellow/model/ProductCategory.java [hunk 15]
```diff
@@ -18,14 +18,14 @@
...
     @Id
-    private String id;
-    private String name;
-    private String desc;
-    private Company companyId;
+    private String Id;
+    private String Name;
+    private String Desc;
+    private Company CompanyId;
     @DBRef
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/model/ProductCategory.java:29]: No need DBRef on AuditData

File: src/main/java/com/yellow/ordermanageryellow/model/ProductCategory.java [hunk 16]
```diff
@@ -18,14 +18,14 @@
...
 
     @Id
-    private String id;
-    private String name;
-    private String desc;
-    private Company companyId;
+    private String Id;
+    private String Name;
+    private String Desc;
+    private Company CompanyId;
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/model/ProductCategory.java:28]: add     @DBRef
on company

**Review by moriyahazan — CHANGES_REQUESTED** [review]
Nice code 
please change according my comments and send again

File: src/main/java/com/yellow/ordermanageryellow/Service/ProductCategoryService.java [hunk 17]
```diff
@@ -1,21 +1,38 @@
...
+            throw new ObjectAlreadyExistException("category name already exist");
+        return this.productCategoryRepository.save(newCategory);
+    }
+
+        public void delete(String categoryId) {
+            this.productCategoryRepository.deleteById(categoryId);
+        }
+
+    public ProductCategory update(ProductCategory updatedCategory){
+            Optional<ProductCategory> Category = this.productCategoryRepository.findById(updatedCategory.get_id());
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Service/ProductCategoryService.java:32]: please update auditdata.updatedate

File: src/main/java/com/yellow/ordermanageryellow/Service/ProductCategoryService.java [hunk 18]
```diff
@@ -1,21 +1,38 @@
 package com.yellow.ordermanageryellow.Service;
+import Exceptions.ObjectAlreadyExistException;
```
Comment (moriyahazan) [inline src/main/java/com/yellow/ordermanageryellow/Service/ProductCategoryService.java:2]: please format all classes

**Review by moriyahazan — APPROVED** [review]
please complete two small comments and merge

**Pull request merged**